src/lib/device/virtual_dev.cpp

```diff
@@ -198,13 +186,19 @@ BlkAllocStatus VirtualDev::do_alloc_blk(blk_count_t nblks, const blk_alloc_hints
 
         do {
             chunk = m_chunk_selector->select_chunk(nblks, hints).get();
-            if (chunk == nullptr) { status = BlkAllocStatus::SPACE_FULL; }
+            if (chunk == nullptr) {
+                status = BlkAllocStatus::SPACE_FULL;
+                break;
+            }
 
-            status = alloc_blk_from_chunk(nblks, hints, out_blkid, chunk);
-            if (status == BlkAllocStatus::SUCCESS || !hints.can_look_for_other_chunk) { break; }
+            status = alloc_blks_from_chunk(nblks, hints, out_blkid, chunk);
+            if ((status == BlkAllocStatus::SUCCESS) || !hints.can_look_for_other_chunk ||
+                (status == BlkAllocStatus::PARTIAL && hints.partial_alloc_ok)) {
+                break;
+            }
         } while (++attempt < m_all_chunks.size());
```

Comment:
It is entirely upto chunk_selector right? Say if have 10 chunks and 9 doesn't have space, I don't want to return no_space since 10th might have it.

Reply:
yes but there is no gareentee in the API contract that it will return N different chunks for N different select_chunk, so  to some point we need to believe chunk_selector does its best scheduling,  for the given allocation_hint. It says no then no (retry just to cover a few edge case that you have add_chunk/delete_blob or similar release space operation in-flight).

Reply:
as discussed in other thread, if we treat HO's selector as an out lier then this makes sense.